src/main/java/org/fogbowcloud/manager/core/threads/ClosedComputeOrdersProcessor.java

```diff
@@ -0,0 +1,110 @@
+    private InstanceProvider remoteProvider;
+
+    private SharedOrderHolders sharedOrderHolders = SharedOrderHolders.getInstance();
+
+    public ClosedComputeOrdersProcessor(Properties properties,InstanceProvider localProvider,
+                                        InstanceProvider remoteProvider) {
+        SharedOrderHolders sharedOrdersHolder = SharedOrderHolders.getInstance();
+        this.closedOrders = sharedOrdersHolder.getClosedOrdersList();
+
+        String closedOrdersSleepTime = properties.getProperty(ConfigurationConstants.CLOSED_ORDERS_SLEEP_TIME_KEY,
+                DefaultConfigurationConstants.CLOSED_ORDERS_SLEEP_TIME);
+        this.sleepTime = Long.valueOf(closedOrdersSleepTime);
+
+        this.localMemberId = properties.getProperty(ConfigurationConstants.XMPP_ID_KEY);
+
+        this.localProvider = localProvider;
+        this.remoteProvider = remoteProvider;
+    }
+
+    @Override
+    public void run() {
+        boolean isActive = true;
+        while (isActive) {
+            try {
+                Order order = this.closedOrders.getNext();
+                if (order != null) {
+                    try {
+                        processClosedOrder(order);
+                        closedOrders.removeItem(order);
+                    } catch (OrderStateTransitionException e) {
+                        LOGGER.error("Error while trying to changing the state of order " + order, e);
+                    }
+                } else {
+                    this.closedOrders.resetPointer();
+                    LOGGER.info(
+                            "There is no closed order to be processed, sleeping for " + this.sleepTime + " milliseconds");
+                    Thread.sleep(this.sleepTime);
+                }
+            } catch (InterruptedException e) {
+                isActive = false;
+                LOGGER.warn("Thread interrupted", e);
+            } catch (Throwable e) {
+                LOGGER.error("Unexpected error", e);
+            }
+        }
+    }
+
+    private void processClosedOrder(Order order) throws OrderStateTransitionException {
+        synchronized (order) {
+            if (order.getOrderInstance() != null &&
+                    order.getOrderInstance().getId() != null) {
+                try {
+                    InstanceProvider provider = getInstanceProviderForOrder(order);
+                    provider.deleteInstance(order.getLocalToken(), order.getOrderInstance());
+
+                    deactivateOrder(order);
+                } catch (Throwable e) {
+                    LOGGER.error(e);
+                }
+            } else {
+                LOGGER.info(String.format("Order %s has no instance associated", order.getId()));
+            }
+        }
+    }
+
+    private void deactivateOrder(Order order) {
+        Map<String, Order> activeOrdersMap = sharedOrderHolders.getActiveOrdersMap();
+        if (activeOrdersMap.get(order.getId()) != null) {
```

Comment:
Is the containsKey method from Map more apropriated?

Reply:
Nice catch! Will fix this in the next commits.